### ISSUE TITLE
SLO Breakage for Kubernetes Pod Startup Latency at scale

### ISSUE THREAD

**Issue by alexkats** [cilium/cilium#22023]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

One of the Kubernetes SLO is the Pod Startup Latency SLO. The current limit is the 99th percentile <= 5s. With the use of cilium Kubernetes is very close to this limit, sometimes even exceeding it.

All tests were performed on GKE's patched Cilium version based on Cilium OSS master. Here's the current status:
* With 100 nodes it didn't go beyond 5 seconds, but comes pretty close from time to time to the limit.
* When running with 500 nodes the average number for P99 is still within 5 seconds, but sometimes goes over and reaches around 5.5 seconds. 
* For 5k nodes the latency is at ~6 seconds on average, but goes even beyond it quite often.

The situation became better with the change #21505, but there're still problems with the latency.

### Cilium Version

master

### Kernel Version

5.10.109

### Kubernetes Version

1.25.2-gke.800

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aanm** [comment]
@alexkats do you have a golang pprof trace output of one of the agents to know which parts are slow?

**Comment by alexkats** [comment]
Hi @aanm, not now, but I'll do more testing, so I'll probably have those later.

**Comment by aanm** [comment]
@alexkats thank you, that will help us know which are the hot paths that we need to fix.

**Comment by ti-mo** [comment]
I've done some local testing around this, I've been experimenting with making ebpf-go's `LoadCollection()` call BPF_PROG_LOAD in parallel.

@alexkats you're describing p99 latency here, but _how_ was that measured? Did you warm up the ELF cache on all nodes by deploying a pod first, taking total pod creation time metric, redeploying the pod, and then taking the delta in total pod creation time? Which metric are you looking at specifically?

If I locally hack up `Endpoint.regenerate()` with a simple `fmt.Println(time.Since(start))`, recreating an existing pod takes around ~50-60ms on a 2-core minikube VM. Of course, if the endpoint's ELF needs to be rebuilt, we're talking ~5-6 seconds, which might explain the p99 outlier you're seeing.

With parallel program loading and sufficient CPU cores to handle the work, I can get this 50ms down to about 30ms. However, the [Kubernetes SLO](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md) currently states: `99th percentile per cluster-day <= 5s`. This is going to heavily depend on the amount of pod creations in any given day. My understanding is, given sufficient pod creations, the 5s <-> 50ms spread should be sufficient to drive down p99 to acceptable levels if taking a 24h average. The ELF cache needs to be given a chance to amortize compile time.

**Comment by marseel** [comment]
FYI, we are redoing testing for PodStartupLatency at 5k nodes scale in https://github.com/cilium/cilium/issues/33778
We utilize the same framework and test that was used by @alexkats 
For additional context, PodStartupLatency measured contains full e2e latency: creating a pod, assigning node by the scheduler, going through the the kubelet, and then Cilium. 
One related issue that affects scheduling latency is https://github.com/cilium/cilium/issues/33462
For 100-node test / 30 pods per node, currently, 99th percentile is ~2.8s
